File: backend/src/main/java/com/backend/application/service/PageTemplateService.java
package com.backend.application.service;

import java.util.List;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import com.backend.application.command.PageTemplateCommands.CreatePageTemplateCommand;
import com.backend.application.command.PageTemplateCommands.CreateTemplateSlotCommand;
import com.backend.application.command.PageTemplateCommands.ReorderTemplateSlotsCommand;
import com.backend.application.command.PageTemplateCommands.UpdatePageTemplateCommand;
import com.backend.application.dto.response.BulkDeleteResultResponse;
import com.backend.application.dto.template.PageTemplateDto;
import com.backend.application.dto.template.TemplateSlotDto;

public interface PageTemplateService {

  List<PageTemplateDto> getAll();

  Page<PageTemplateDto> getTemplates(Pageable pageable, String search);

  List<PageTemplateDto> getActiveTemplates();

  PageTemplateDto getById(Long id);

  PageTemplateDto create(CreatePageTemplateCommand command);

  PageTemplateDto update(Long id, UpdatePageTemplateCommand command);

  void delete(Long id);

  BulkDeleteResultResponse bulkDeletePageTemplates(List<Long> ids);

  TemplateSlotDto addSlot(Long templateId, CreateTemplateSlotCommand command);

  void removeSlot(Long templateId, String slotName);

  void reorderSlots(Long templateId, ReorderTemplateSlotsCommand command);

  void assignTemplateToPage(Long pageId, Long templateId);
}

File: storefront/src/app/modules/admin/custom/templates/page-template.service.ts
import { Injectable } from '@angular/core';
import { ApiResponse, CrudHttpService } from '@core/crud';
import { map, Observable, take } from 'rxjs';
import {
    CreatePageTemplateDto,
    CreateTemplateSlotDto,
    PageTemplate,
    TemplateSlot,
    UpdatePageTemplateDto
} from './page-template.types';

@Injectable({ providedIn: 'root' })
export class PageTemplateService extends CrudHttpService<PageTemplate, CreatePageTemplateDto, UpdatePageTemplateDto> {
    protected endpoints = {
        list: 'pageTemplates' as const,
        getById: 'pageTemplateById' as const,
        create: 'pageTemplates' as const,
        update: 'pageTemplateById' as const,
        delete: 'pageTemplateById' as const
    };

    getActive(): Observable<PageTemplate[]> {
        return this.api
            .get<ApiResponse<PageTemplate[]>>('pageTemplatesActive')
            .pipe(
                take(1),
                map(response => response.data)
            );
    }

    addSlot(templateId: number, dto: CreateTemplateSlotDto): Observable<TemplateSlot> {
        return this.api
            .post<ApiResponse<TemplateSlot>>('pageTemplateSlots', dto, { id: templateId })
            .pipe(
                take(1),
                map(response => response.data)
            );
    }

    removeSlot(templateId: number, slotName: string): Observable<void> {
        return this.api
            .delete<ApiResponse<void>>('pageTemplateSlot', { id: templateId, slotName })
            .pipe(
                take(1),
                map(() => void 0)
            );
    }

    assignToPage(templateId: number, pageId: number): Observable<void> {
        return this.api
            .post<ApiResponse<void>>('pageTemplateAssign', {}, { id: templateId, pageId })
            .pipe(
                take(1),
                map(() => void 0)
            );
    }

    reorderSlots(templateId: number, slotNames: string[]): Observable<void> {
        return this.api
            .put<ApiResponse<void>>('pageTemplateSlotsReorder', { items: slotNames }, { id: templateId })
            .pipe(
                take(1),
                map(() => void 0)
            );
    }
}
